pkg/streaming/util/message/adaptor/message.go

```diff
@@ -0,0 +1,173 @@
+	return &msgstream.MsgPack{
+		BeginTs:        allTsMsgs[0].BeginTs(),
+		EndTs:          allTsMsgs[len(allTsMsgs)-1].EndTs(),
+		Msgs:           allTsMsgs,
+		StartPositions: []*msgstream.MsgPosition{allTsMsgs[0].Position()},
+		EndPositions:   []*msgstream.MsgPosition{allTsMsgs[len(allTsMsgs)-1].Position()},
+	}, nil
+}
+
+func fromMessageToTsMsgVOld(msg message.ImmutableMessage) ([]msgstream.TsMsg, error) {
+	panic("Not implemented")
+}
+
+// fromMessageToTsMsgV1 converts message to ts message.
+func fromMessageToTsMsgV1(msg message.ImmutableMessage) ([]msgstream.TsMsg, error) {
+	tsMsg, err := unmashalerDispatcher.Unmarshal(msg.Payload(), commonpb.MsgType(msg.MessageType()))
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to unmarshal message")
+	}
+	tsMsg.SetTs(msg.TimeTick())
+	tsMsg.SetPosition(&msgpb.MsgPosition{
+		ChannelName: msg.VChannel(),
+		// from the last confirmed message id, you can read all messages which timetick is greater or equal than current message id.
+		MsgID:     MustGetMQWrapperIDFromMessage(msg.LastConfirmedMessageID()).Serialize(),
+		MsgGroup:  "", // Not important any more.
+		Timestamp: msg.TimeTick(),
+	})
+
+	return recoverMessageFromHeader(tsMsg, msg)
+}
+
+// recoverMessageFromHeader recovers message from header.
+func recoverMessageFromHeader(tsMsg msgstream.TsMsg, msg message.ImmutableMessage) ([]msgstream.TsMsg, error) {
+	switch msg.MessageType() {
+	case message.MessageTypeInsert:
+		insertMessage, err := message.AsImmutableInsertMessage(msg)
+		if err != nil {
+			return nil, errors.Wrap(err, "Failed to convert message to insert message")
+		}
+		// insertMsg has multiple partition and segment assignment is done by insert message header.
+		// so recover insert message from header before send it.
+		return recoverInsertMsgFromHeader(tsMsg.(*msgstream.InsertMsg), insertMessage.MessageHeader(), msg.TimeTick())
+	default:
+		return []msgstream.TsMsg{tsMsg}, nil
+	}
+}
+
+// recoverInsertMsgFromHeader recovers insert message from header.
+func recoverInsertMsgFromHeader(insertMsg *msgstream.InsertMsg, header *message.InsertMessageHeader, timetick uint64) ([]msgstream.TsMsg, error) {
+	if insertMsg.GetCollectionID() != header.GetCollectionId() {
+		panic("unreachable code, collection id is not equal")
+	}
+	// header promise a batch insert on vchannel in future, so header has multiple partition.
+	var assignments *message.PartitionSegmentAssignment
+	for _, p := range header.Partitions {
+		if p.GetPartitionId() == insertMsg.GetPartitionID() {
+			assignments = p
+			break
+		}
+	}
+	if assignments == nil {
+		panic("unreachable code, partition id is not exist")
+	}
+	tsMsgs := make([]msgstream.TsMsg, 0, len(assignments.SegmentAssignments))
+	// timetick should has been assign at streaming node.
+	// so overwrite the timetick on insertRequest.
+
+	left := 0
+	right := 0
+	for _, assignment := range assignments.SegmentAssignments {
+		right = left + int(assignment.Row)
```

Comment:
Wrap an independent method to build `tsMsgs`  slice for a `segmentAssignment`.

Reply:
ok, fix it soon